Using reflection-free serializers breaks in 3.35

### Describe the bug

I have an app that uses the spring compatibility layer and also configures to use the reflection free serializers. When upgrading from 3.34 to 3.35 the app fails.

### Expected behavior

I would expect the upgrade to go smoothly.

### Actual behavior

I get this error:

```
java.lang.RuntimeException: java.lang.ExceptionInInitializerError

	at io.quarkus.test.junit.QuarkusTestExtension.throwBootFailureException(QuarkusTestExtension.java:668)
	at io.quarkus.test.junit.QuarkusTestExtension.interceptTestClassConstructor(QuarkusTestExtension.java:763)
	at java.base/java.util.Optional.orElseGet(Optional.java:364)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
Caused by: java.lang.ExceptionInInitializerError
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:536)
	at java.base/java.lang.Class.forName(Class.java:515)
	at io.quarkus.runner.bootstrap.StartupActionImpl.run(StartupActionImpl.java:369)
	at io.quarkus.test.junit.QuarkusTestExtension.doJavaStart(QuarkusTestExtension.java:232)
	at io.quarkus.test.junit.QuarkusTestExtension.ensureStarted(QuarkusTestExtension.java:642)
	at io.quarkus.test.junit.QuarkusTestExtension.beforeAll(QuarkusTestExtension.java:687)
	... 1 more
Caused by: java.lang.RuntimeException: Failed to start quarkus
	at io.quarkus.runner.ApplicationImpl.<clinit>(Unknown Source)
	... 8 more
Caused by: java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at io.quarkus.resteasy.reactive.jackson.runtime.serialisers.GeneratedSerializersRegister.addSerializer(GeneratedSerializersRegister.java:44)
	at io.quarkus.resteasy.reactive.jackson.runtime.ResteasyReactiveServerJacksonRecorder.recordGeneratedSerializer(ResteasyReactiveServerJacksonRecorder.java:85)
	at io.quarkus.runner.recorded.ResteasyReactiveJacksonProcessor$handleEndpointParams2008319968.deploy_0(Unknown Source)
	at io.quarkus.runner.recorded.ResteasyReactiveJacksonProcessor$handleEndpointParams2008319968.deploy(Unknown Source)
	... 9 more
Caused by: java.lang.reflect.InvocationTargetException
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:502)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:486)
	at io.quarkus.resteasy.reactive.jackson.runtime.serialisers.GeneratedSerializersRegister.addSerializer(GeneratedSerializersRegister.java:40)
	... 12 more
Caused by: java.lang.IllegalAccessError: failed to access class org.springframework.http.DefaultHttpStatusCode from class org.springframework.http.DefaultHttpStatusCode$quarkusjacksonserializer (org.springframework.http.DefaultHttpStatusCode is in unnamed module of loader io.quarkus.bootstrap.classloading.QuarkusClassLoader @6b9c69a9; org.springframework.http.DefaultHttpStatusCode$quarkusjacksonserializer is in unnamed module of loader io.quarkus.bootstrap.classloading.QuarkusClassLoader @563172d3)
	at org.springframework.http.DefaultHttpStatusCode$quarkusjacksonserializer.<init>(Unknown Source)
	... 15 more
```

### How to Reproduce?

App: https://github.com/quarkusio/spring-quarkus-perf-comparison/tree/main/quarkus3-spring-compatibility

Steps to reproduce:
1. Clone https://github.com/quarkusio/spring-quarkus-perf-comparison.git
2. `cd quarkus3-spring-compatibility`
4. `./mvnw clean test -Dtest=FruitControllerEndToEndTest -Dquarkus.platform.version=3.35.2`
### Output of `uname -a` or `ver`

`Darwin edeandrea-m4max 25.4.0 Darwin Kernel Version 25.4.0: Thu Mar 19 19:33:25 PDT 2026; root:xnu-12377.101.15~1/RELEASE_ARM64_T6041 arm64`

### Output of `java -version`

```
openjdk version "21.0.10" 2026-01-20 LTS
OpenJDK Runtime Environment Temurin-21.0.10+7 (build 21.0.10+7-LTS)
OpenJDK 64-Bit Server VM Temurin-21.0.10+7 (build 21.0.10+7-LTS, mixed mode, sharing)
```

### Quarkus version or git rev

`3.35.2`

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

```
Apache Maven 3.9.11 (3e54c93a704957b63ee3494413a2b544fd3d825b)
Maven home: /Users/edeandre/.m2/wrapper/dists/apache-maven-3.9.11/a2d47e15
Java version: 21.0.10, vendor: Eclipse Adoptium, runtime: /Users/edeandre/.sdkman/candidates/java/21.0.10-tem
Default locale: en_US, platform encoding: UTF-8
OS name: "mac os x", version: "26.4.1", arch: "aarch64", family: "mac"
```

### Additional information

_No response_

Hi @SubhamKrGuptaDev I still don't think you are understanding. Switching the version of the quarkus cli isn't the correct thing. Changing the quarkus version in `pom.xml` introduces the problem:

```
 ./mvnw clean test -Dtest=FruitControllerEndToEndTest -Dquarkus.platform.version=3.35.2

[INFO] --- surefire:3.5.5:test (default-test) @ quarkus3-spring-compatibility ---
[INFO] Using auto detected provider org.apache.maven.surefire.junitplatform.JUnitPlatformProvider
[INFO] 
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[ERROR] Tests run: 4, Failures: 0, Errors: 1, Skipped: 3, Time elapsed: 1.483 s <<< FAILURE! -- in org.acme.e2e.FruitControllerEndToEndTest
[ERROR] org.acme.e2e.FruitControllerEndToEndTest.getAll -- Time elapsed: 0.004 s <<< ERROR!
java.lang.RuntimeException: java.lang.ExceptionInInitializerError
	at io.quarkus.test.junit.QuarkusTestExtension.throwBootFailureException(QuarkusTestExtension.java:668)
	at io.quarkus.test.junit.QuarkusTestExtension.interceptTestClassConstructor(QuarkusTestExtension.java:763)
	at java.base/java.util.Optional.orElseGet(Optional.java:364)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1604)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1604)
Caused by: java.lang.ExceptionInInitializerError
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:547)
	at io.quarkus.runner.bootstrap.StartupActionImpl.run(StartupActionImpl.java:369)
	at io.quarkus.test.junit.QuarkusTestExtension.doJavaStart(QuarkusTestExtension.java:232)
	at io.quarkus.test.junit.QuarkusTestExtension.ensureStarted(QuarkusTestExtension.java:642)
	at io.quarkus.test.junit.QuarkusTestExtension.beforeAll(QuarkusTestExtension.java:687)
	... 1 more
Caused by: java.lang.RuntimeException: Failed to start quarkus
	at io.quarkus.runner.ApplicationImpl.<clinit>(Unknown Source)
	... 7 more
Caused by: java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at io.quarkus.resteasy.reactive.jackson.runtime.serialisers.GeneratedSerializersRegister.addSerializer(GeneratedSerializersRegister.java:44)
	at io.quarkus.resteasy.reactive.jackson.runtime.ResteasyReactiveServerJacksonRecorder.recordGeneratedSerializer(ResteasyReactiveServerJacksonRecorder.java:85)
	at io.quarkus.runner.recorded.ResteasyReactiveJacksonProcessor$handleEndpointParams2008319968.deploy_0(Unknown Source)
	at io.quarkus.runner.recorded.ResteasyReactiveJacksonProcessor$handleEndpointParams2008319968.deploy(Unknown Source)
	... 8 more
Caused by: java.lang.reflect.InvocationTargetException
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:483)
	at io.quarkus.resteasy.reactive.jackson.runtime.serialisers.GeneratedSerializersRegister.addSerializer(GeneratedSerializersRegister.java:40)
	... 11 more
Caused by: java.lang.IllegalAccessError: failed to access class org.springframework.http.DefaultHttpStatusCode from class org.springframework.http.DefaultHttpStatusCode$quarkusjacksonserializer (org.springframework.http.DefaultHttpStatusCode is in unnamed module of loader io.quarkus.bootstrap.classloading.QuarkusClassLoader @15515c51; org.springframework.http.DefaultHttpStatusCode$quarkusjacksonserializer is in unnamed module of loader io.quarkus.bootstrap.classloading.QuarkusClassLoader @27bb74e1)
	at org.springframework.http.DefaultHttpStatusCode$quarkusjacksonserializer.<init>(Unknown Source)
	... 14 more

[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Errors: 
[ERROR]   FruitControllerEndToEndTest.getAll » Runtime java.lang.ExceptionInInitializerError
[INFO] 
[ERROR] Tests run: 4, Failures: 0, Errors: 1, Skipped: 3

```

--- a/extensions/resteasy-reactive/rest-jackson/deployment/src/main/java/io/quarkus/resteasy/reactive/jackson/deployment/processor/JacksonCodeGenerator.java
+++ b/extensions/resteasy-reactive/rest-jackson/deployment/src/main/java/io/quarkus/resteasy/reactive/jackson/deployment/processor/JacksonCodeGenerator.java
@@ -1,5 +1,7 @@
 package io.quarkus.resteasy.reactive.jackson.deployment.processor;
 
+import static io.quarkus.bootstrap.classloading.QuarkusClassLoader.isApplicationClass;
+
 import java.lang.reflect.Modifier;
 import java.util.ArrayDeque;
 import java.util.ArrayList;
@@ -313,7 +315,7 @@ public abstract class JacksonCodeGenerator {
 
     private void registerTypeToBeGenerated(String typeName) {
         ClassInfo classInfo = jandexIndex.getClassByName(typeName);
-        if (classInfo == null) {
+        if (classInfo == null || !isRuntimeAccessible(classInfo, typeName)) {
             return;
         }
         if (vetoedClass(classInfo, typeName)) {
@@ -327,6 +329,10 @@ public abstract class JacksonCodeGenerator {
         if (shouldGenerateCodeFor(classInfo)) {
             toBeGenerated.add(classInfo);
         }
+    }
+
+    private static boolean isRuntimeAccessible(ClassInfo classInfo, String className) {
+        return Modifier.isPublic(classInfo.flags()) || isApplicationClass(className);
     }
 
     protected boolean shouldGenerateCodeFor(ClassInfo classInfo) {

--- a/extensions/resteasy-reactive/rest-jackson/deployment/src/test/java/io/quarkus/resteasy/reactive/jackson/deployment/test/generated/AbstractGeneratedAnnotationTest.java
+++ b/extensions/resteasy-reactive/rest-jackson/deployment/src/test/java/io/quarkus/resteasy/reactive/jackson/deployment/test/generated/AbstractGeneratedAnnotationTest.java
@@ -686,6 +686,32 @@ public abstract class AbstractGeneratedAnnotationTest {
         assertTrue(alphaPos < middlePos, "alpha should appear before middle");
     }
 
+    // --- Package-protected class with single int constructor ---
+
+    @Test
+    public void testPackageProtectedSerialization() {
+        RestAssured.get("/generated/package-protected")
+                .then()
+                .statusCode(200)
+                .contentType("application/json")
+                .body("value", Matchers.is(42))
+                .body("label", Matchers.is("custom-label"));
+    }
+
+    @Test
+    public void testPackageProtectedDeserialization() {
+        given()
+                .contentType("application/json")
+                .body("{\"value\":7,\"label\":\"from-json\"}")
+                .when()
+                .post("/generated/package-protected")
+                .then()
+                .statusCode(200)
+                .contentType("application/json")
+                .body("value", Matchers.is(7))
+                .body("label", Matchers.is("from-json"));
+    }
+
     // --- @JsonRawValue ---
 
     @Test

--- a/extensions/resteasy-reactive/rest-jackson/deployment/src/test/java/io/quarkus/resteasy/reactive/jackson/deployment/test/generated/GeneratedAnnotationResource.java
+++ b/extensions/resteasy-reactive/rest-jackson/deployment/src/test/java/io/quarkus/resteasy/reactive/jackson/deployment/test/generated/GeneratedAnnotationResource.java
@@ -403,4 +403,21 @@ public class GeneratedAnnotationResource {
         bean.setRawJson("{\"nested\":\"value\",\"count\":1}");
         return bean;
     }
+
+    // --- PackageProtectedBean: package-protected class with single int constructor ---
+
+    @GET
+    @Path("/package-protected")
+    public PackageProtectedBean getPackageProtected() {
+        PackageProtectedBean bean = new PackageProtectedBean(42);
+        bean.setLabel("custom-label");
+        return bean;
+    }
+
+    @POST
+    @Path("/package-protected")
+    @Consumes(MediaType.APPLICATION_JSON)
+    public PackageProtectedBean echoPackageProtected(PackageProtectedBean bean) {
+        return bean;
+    }
 }

--- a/extensions/resteasy-reactive/rest-jackson/deployment/src/test/java/io/quarkus/resteasy/reactive/jackson/deployment/test/generated/GeneratedAnnotationTest.java
+++ b/extensions/resteasy-reactive/rest-jackson/deployment/src/test/java/io/quarkus/resteasy/reactive/jackson/deployment/test/generated/GeneratedAnnotationTest.java
@@ -47,7 +47,8 @@ public class GeneratedAnnotationTest extends AbstractGeneratedAnnotationTest {
                                     IgnoreTypeBean.class,
                                     IncludeBean.class,
                                     PropertyOrderBean.class,
-                                    RawValueBean.class)
+                                    RawValueBean.class,
+                                    PackageProtectedBean.class)
                             .addAsResource(new StringAsset(
                                     "quarkus.jackson.fail-on-unknown-properties=true\n" +
                                             "quarkus.rest.jackson.optimization.enable-reflection-free-serializers=false\n"),

--- a/extensions/resteasy-reactive/rest-jackson/deployment/src/test/java/io/quarkus/resteasy/reactive/jackson/deployment/test/generated/GeneratedAnnotationWithReflectionFreeSerializersTest.java
+++ b/extensions/resteasy-reactive/rest-jackson/deployment/src/test/java/io/quarkus/resteasy/reactive/jackson/deployment/test/generated/GeneratedAnnotationWithReflectionFreeSerializersTest.java
@@ -50,7 +50,8 @@ public class GeneratedAnnotationWithReflectionFreeSerializersTest extends Abstra
                                     IgnoreTypeBean.class,
                                     IncludeBean.class,
                                     PropertyOrderBean.class,
-                                    RawValueBean.class)
+                                    RawValueBean.class,
+                                    PackageProtectedBean.class)
                             .addAsResource(new StringAsset(
                                     "quarkus.jackson.fail-on-unknown-properties=true\n" +
                                             "quarkus.rest.jackson.optimization.enable-reflection-free-serializers=true\n"),

--- a/extensions/resteasy-reactive/rest-jackson/deployment/src/test/java/io/quarkus/resteasy/reactive/jackson/deployment/test/generated/PackageProtectedBean.java
+++ b/extensions/resteasy-reactive/rest-jackson/deployment/src/test/java/io/quarkus/resteasy/reactive/jackson/deployment/test/generated/PackageProtectedBean.java
@@ -1,0 +1,24 @@
+package io.quarkus.resteasy.reactive.jackson.deployment.test.generated;
+
+final class PackageProtectedBean {
+
+    private final int value;
+    private String label;
+
+    public PackageProtectedBean(int value) {
+        this.value = value;
+        this.label = "item-" + value;
+    }
+
+    public int getValue() {
+        return value;
+    }
+
+    public String getLabel() {
+        return label;
+    }
+
+    public void setLabel(String label) {
+        this.label = label;
+    }
+}

